public/app/features/alerting/unified/rule-list/hooks/grafanaFilter.ts

```diff
@@ -0,0 +1,118 @@
+import { PromRuleDTO, PromRuleGroupDTO } from 'app/types/unified-alerting-dto';
+
+import { GrafanaPromRulesOptions } from '../../api/prometheusApi';
+import { shouldUseBackendFilters } from '../../featureToggles';
+import { RulesFilter } from '../../search/rulesSearchParser';
+
+import { buildTitleSearch, normalizeFilterState } from './filterNormalization';
+import {
+  GroupFilterConfig,
+  RuleFilterConfig,
+  dashboardUidFilter,
+  dataSourceNamesFilter,
+  freeFormFilter,
+  groupMatches,
+  groupNameFilter,
+  labelsFilter,
+  namespaceFilter,
+  pluginsFilter,
+  ruleMatches,
+  ruleNameFilter,
+  ruleTypeFilter,
+} from './filterPredicates';
+
+/**
+ * Determines if client-side filtering is needed for Grafana-managed rules.
+ */
+export function hasClientSideFilters(filterState: RulesFilter): boolean {
+  const { ruleFilterConfig: grafanaFilterProcessingConfig, groupFilterConfig: grafanaGroupFilterConfig } =
+    buildGrafanaFilterConfigs();
+
+  // Check each rule filter: if the config has a non-null handler AND the filter state has a value, we need client-side filtering
+  const hasActiveRuleFilters =
+    (grafanaFilterProcessingConfig.freeFormWords !== null && filterState.freeFormWords.length > 0) ||
+    (grafanaFilterProcessingConfig.ruleName !== null && Boolean(filterState.ruleName)) ||
+    (grafanaFilterProcessingConfig.ruleState !== null && Boolean(filterState.ruleState)) ||
+    (grafanaFilterProcessingConfig.ruleType !== null && Boolean(filterState.ruleType)) ||
+    (grafanaFilterProcessingConfig.dataSourceNames !== null && filterState.dataSourceNames.length > 0) ||
+    (grafanaFilterProcessingConfig.labels !== null && filterState.labels.length > 0) ||
+    (grafanaFilterProcessingConfig.ruleHealth !== null && Boolean(filterState.ruleHealth)) ||
+    (grafanaFilterProcessingConfig.dashboardUid !== null && Boolean(filterState.dashboardUid)) ||
+    (grafanaFilterProcessingConfig.plugins !== null && Boolean(filterState.plugins)) ||
+    (grafanaFilterProcessingConfig.contactPoint !== null && Boolean(filterState.contactPoint));
+
+  // Check each group filter: if the config has a non-null handler AND the filter state has a value, we need client-side filtering
+  const hasActiveGroupFilters =
+    (grafanaGroupFilterConfig.namespace !== null && Boolean(filterState.namespace)) ||
+    (grafanaGroupFilterConfig.groupName !== null && Boolean(filterState.groupName));
+
+  return hasActiveRuleFilters || hasActiveGroupFilters;
+}
```

Comment:
This is the main change made in this PR. I wanted `hasClientSideFilters` to use `ruleFilterConfig` and `groupFilterConfig` so we only build the config once and the use it in other places